### DISCUSSION TITLE
Getting a value from an Option without 'Some' in value

### DISCUSSION THREAD

**Discussion by speterson-zoll** [louthy/language-ext#1142] [Q&A] [answered]
I have a function that takes in two `DateTime` parameters and turns them into `Option<DateTime>`:

```csharp
public static Validation<Error, bool> ValidateDateRange(Option<DateTime> startDateTime,
      Option<DateTime> endDateTime)
    {
      var result = map(Tuple(startDateTime, endDateTime), (start, end) => (start.IsNone || end.IsNone) || start <= end);

      return result
        ? Success<Error, bool>(result)
        : Fail<Error, bool>(Error.New($"The value of the start date '{startDateTime.GetValue()}' cannot be greater than the end date '{endDateTime.GetValue()}'."));
    }
```

As you can see, I created an extension method called `GetValue` based on something I read in another discussion.  I made it generic and it looks like this:

```csharp
public static class GetValues
{
  public static string? GetValue<T>(this Option<T> input) =>
    input.Match(
      Some: x => x!.ToString(),
      None: () => string.Empty);
}
```

My goal is to include the value of the `DateTime` in my error message, but I can't seem to get the actual value out with a call to `Map`.  For example, this call: `startDateTime.Map(x => x)` returns `Some(11/15/2022 8:06:36 AM)`.  How do I get rid of `Some`?  The variable `x` in the `Map` call shows up as a `DateTime` object, so why am I still getting `Some` in my value?

**Comment by louthy** [comment, marked as answer]
Not sure if I'm entirely parsing what you're asking, but I think there's an alternative route that's more elegant.  Really you want to work with the `Some` state when you know it is a `Some` state, not try to hack it to be either the value or an empty string after-the-fact:

```c#
public static Error InvalidRange(DateTime s, DateTime e) =>
    Error.New($"The value of the start date '{s}' cannot be greater than the end date '{e}'.");

public static Validation<Error, Unit> ValidateDateRange(
    Option<DateTime> startDateTime,
    Option<DateTime> endDateTime) =>
    (startDateTime.Case, endDateTime.Case) switch
    {
        (DateTime s, DateTime e) when s > e => InvalidRange(s, e),
        _ => Success<Error, Unit>(unit),
    };
```
What this does is it pattern-matches the `Option` _when_ the `start` is greater than the `end`.  This is the only failure state I can gather from your example.  So, as soon as we know it's wrong, we fail.  The subsequent catch of any other value returns `Success`.

I have also changed it to return `unit` rather than `bool`, because in your example I believe you will always have `true` if the validation is successful.  Which means it's a single value, i.e. a `unit`, not a `bool`.

This test:
```c#
System.Console.WriteLine(ValidateDateRange(DateTime.Now, DateTime.Now.AddDays(1)));
System.Console.WriteLine(ValidateDateRange(DateTime.Now, DateTime.Now.AddDays(-1)));
System.Console.WriteLine(ValidateDateRange(None, DateTime.Now));
System.Console.WriteLine(ValidateDateRange(DateTime.Now, None));
```
Yields:
```c#
Success(())
Fail([The value of the start date '15/11/2022 17:17:04' cannot be greater than the end date '14/11/2022 17:17:04'.])
Success(())
Success(())
```

**Reply by speterson-zoll** [reply]
Thanks you once again.  I had not made use of the `Case` property.
